presto-orc/src/main/java/com/facebook/presto/orc/BaseOrcRecordReader.java

```diff
@@ -0,0 +1,610 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc;
+
+import com.facebook.presto.memory.context.AggregatedMemoryContext;
+import com.facebook.presto.orc.metadata.ColumnEncoding;
+import com.facebook.presto.orc.metadata.MetadataReader;
+import com.facebook.presto.orc.metadata.OrcType;
+import com.facebook.presto.orc.metadata.PostScript;
+import com.facebook.presto.orc.metadata.StripeInformation;
+import com.facebook.presto.orc.metadata.statistics.ColumnStatistics;
+import com.facebook.presto.orc.metadata.statistics.StripeStatistics;
+import com.facebook.presto.orc.reader.BaseStreamReader;
+import com.facebook.presto.orc.stream.InputStreamSources;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Predicate;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Maps;
+import com.google.common.io.Closer;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import io.airlift.units.DataSize;
+import org.joda.time.DateTimeZone;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.orc.BaseOrcRecordReader.LinearProbeRangeFinder.createTinyStripesRangeFinder;
+import static com.facebook.presto.orc.OrcDataSourceUtils.mergeAdjacentDiskRanges;
+import static com.facebook.presto.orc.OrcReader.MAX_BATCH_SIZE;
+import static com.facebook.presto.orc.OrcWriteValidation.WriteChecksumBuilder.createWriteChecksumBuilder;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static java.lang.Math.toIntExact;
+import static java.util.Comparator.comparingLong;
+import static java.util.Objects.requireNonNull;
+
+abstract class BaseOrcRecordReader
+        implements Closeable
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(BaseOrcRecordReader.class).instanceSize();
+
+    private final OrcDataSource orcDataSource;
+
+    protected final BaseStreamReader[] baseStreamReaders;
```

Comment:
Interleaved protected and private variables kinda make the class a bit less abstracted. I put some comments to address the issues but not all. Hope we could keep all variables private. @wenleix, @tdcmeehan, @yingsu00 could possibly provide more ideas.

Reply:
@highker James, thanks for the suggestion about `prepareNextBatch()`. I applied it along with a handful of other tricks and removed all `protected` member variables from `AbstractOrcRecorddReader`.